File: frontend/src/stimulus/controllers/dynamic/backlogs/backlog.ts
//-- copyright
// OpenProject is an open source project management software.
// Copyright (C) the OpenProject GmbH
//
// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License version 3.
//
// OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
// Copyright (C) 2006-2013 Jean-Philippe Lang
// Copyright (C) 2010-2013 the ChiliProject Team
//
// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License
// as published by the Free Software Foundation; either version 2
// of the License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
//
// See COPYRIGHT and LICENSE files for more details.
//++

import { Burndown } from './burndown';
import { EditableSprint } from './sprint';
import { EditableStory, EditableStoryType } from './story';

/******************************************
  BACKLOG
  A backlog is a visual representation of
  a sprint and its stories. It is not a
  sprint. Imagine it this way: A sprint is
  a start and end date, and a set of
  objectives. A backlog is something you
  would draw up on the board or a spread-
  sheet (or in Redmine Backlogs!) to
  visualize the sprint.
******************************************/
export class Backlog {
  private $:JQuery;
  private el:HTMLElement;
  burndown:Burndown;

  constructor(el:HTMLElement) {
    this.$ = $(el);
    this.el = el;

    // Associate this object with the element for later retrieval
    this.$.data('this', this);

    // Make the list sortable
    this.getList().sortable({
      connectWith: '.stories',
      dropOnEmpty: true,
      start: this.dragStart,
      stop: this.dragStop,
      update: this.dragComplete,
      receive: this.dragChanged,
      remove: this.dragChanged,
      containment: $('#backlogs_container'),
      cancel: 'input, textarea, button, select, option, .prevent_drag',
      scroll: true,
      helper(ui) {
        const $clone = $(ui).clone();
        $clone.css('position', 'absolute');
        return $clone.get(0) as unknown as Element;
      },
    });

    // Observe menu items
    this.$.find('.add_new_story').click((e) => this.handleNewStoryClick(e));

    if (this.isSprintBacklog()) {
      new EditableSprint(this.getSprint()[0]);
      this.burndown = new Burndown(this.$.find('.show_burndown_chart')[0]);
      this.burndown.setSprintId(this.getSprint().data('this').getID());
    }

    // Initialize each item in the backlog
    this.getStories().each((index, element) => {
      // refers to an element with class="story"
      new EditableStory(element);
    });

    if (this.isSprintBacklog()) {
      this.refresh();
    }
  }

  dragChanged(e:JQueryEventObject, ui:JQueryUI.SortableUIParams) {
    $(this).parents('.backlog').data('this').refresh();
  }

  dragComplete(e:JQueryEventObject, ui:JQueryUI.SortableUIParams) {
    const isDropTarget = (ui.sender === null || ui.sender === undefined);

    // jQuery triggers dragComplete of source and target.
    // Thus we have to check here. Otherwise, the story
    // would be saved twice.
    if (isDropTarget) {
      ui.item.data('this').saveDragResult();
    }
  }

  dragStart(e:JQueryEventObject, ui:JQueryUI.SortableUIParams) {
    ui.item.addClass('dragging');
  }

  dragStop(e:JQueryEventObject, ui:JQueryUI.SortableUIParams) {
    ui.item.removeClass('dragging');
  }

  getSprint() {
    return $(this.el).find('.model.sprint').first();
  }

  getStories() {
    return this.getList().children('.story');
  }

  getList() {
    return this.$.children('.stories').first();
  }

  handleNewStoryClick(e:JQuery.TriggeredEvent) {
    const toggler = $(e.currentTarget).parents('.header').find('.toggler');
    if (toggler.hasClass('closed')) {
      toggler.click();
    }
    e.preventDefault();
    this.newStory();
  }

  // return true if backlog has an element with class="sprint"
  isSprintBacklog() {
    return $(this.el).find('.sprint').length === 1;
  }

  newStory() {
    const story = $('#story_template').children().first().clone();
    this.getList().prepend(story);

    const o = new EditableStory(story[0]);
    o.edit();

    story.find('.editor').first().focus();
  }

  refresh() {
    this.recalcVelocity();
    this.recalcOddity();
  }

  recalcVelocity() {
    let total:any;

    if (!this.isSprintBacklog()) {
      return;
    }

    total = 0;
    this.getStories().each((index, element) => {
      total += ($(element).data('this') as EditableStoryType).getPoints();
    });
    this.$.children('.header').children('.velocity').text(total);
  }

  recalcOddity() {
    this.$.find('.story:even').removeClass('odd').addClass('even');
    this.$.find('.story:odd').removeClass('even').addClass('odd');
  }
}

File: modules/backlogs/app/models/backlog.rb
#-- copyright
# OpenProject is an open source project management software.
# Copyright (C) the OpenProject GmbH
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License version 3.
#
# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
# Copyright (C) 2006-2013 Jean-Philippe Lang
# Copyright (C) 2010-2013 the ChiliProject Team
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
#
# See COPYRIGHT and LICENSE files for more details.
#++

class Backlog
  extend ActiveModel::Naming

  attr_accessor :sprint, :stories

  delegate :id, to: :sprint, prefix: true

  def self.for(sprint:, project:)
    owner_backlog = sprint.settings(project)&.display == VersionSetting::DISPLAY_RIGHT
    stories, truncated = Story.backlog_for(project.id, sprint.id)
    new(sprint:, stories:, owner_backlog:, truncated:)
  end

  def self.owner_backlogs(project)
    # Loop one bounded query per backlog instead of a single unbounded query
    # across all of them. For typical projects (low single digits of backlog
    # columns) the additional round trips are cheap and the LIMIT keeps any
    # one column from dominating page-render time.
    backlogs = Sprint.apply_to(project).with_status_open.displayed_right(project).order(:name)
    backlogs.map do |sprint|
      stories, truncated = Story.backlog_for(project.id, sprint.id)
      new(stories:, owner_backlog: true, sprint:, truncated:)
    end
  end

  def self.sprint_backlogs(project)
    sprints = Sprint.apply_to(project).with_status_open.displayed_left(project).order_by_date
    sprints.map do |sprint|
      stories, truncated = Story.backlog_for(project.id, sprint.id)
      new(stories:, sprint:, truncated:)
    end
  end

  def self.inbox_backlog(project, include_closed: false)
    stories, truncated = Story.inbox_for(project.id, include_closed:)
    new(sprint: nil, stories:, inbox: true, truncated:)
  end

  def initialize(sprint:, stories:, owner_backlog: false, inbox: false, truncated: false)
    @sprint = sprint
    @stories = stories
    @owner_backlog = owner_backlog
    @inbox = inbox
    @truncated = truncated
  end

  def updated_at
    @stories.max_by(&:updated_at).try(:updated_at)
  end

  def owner_backlog?
    !!@owner_backlog
  end

  def sprint_backlog?
    !owner_backlog? && !inbox?
  end

  def inbox?
    !!@inbox
  end

  def truncated?
    !!@truncated
  end

  def column_limit
    Story::COLUMN_LIMIT
  end

  def to_key
    [inbox? ? "inbox" : sprint_id]
  end
end
